### PR TITLE
[WIP] Piece Rangers should connect lazily to storage nodes

### PR DESCRIPTION
**Pull request by mniewrzal**
This change is [Reviewable](https://reviewable.io/reviews/storj/storj/314)

### CODE REVIEW

File: pkg/piecestore/rpc/client/client.go [hunk 1]
```diff
@@ -64,8 +68,10 @@ func NewPSClient(conn *grpc.ClientConn, bandwidthMsgSize int, prikey crypto.Priv
 	}
 
 	return &Client{
-		conn:             conn,
-		route:            pb.NewPieceStoreRoutesClient(conn),
+		conn:             nil,
+		route:            nil,
```
Comment (kaloyan-raev) [inline pkg/piecestore/rpc/client/client.go]: You can just remove the above two lines. The default vaue for these fields is `nil` if you miss setting them.

File: pkg/piecestore/rpc/client/client.go [hunk 2]
```diff
@@ -166,3 +179,16 @@ func (client *Client) sign(msg []byte) (signature []byte, err error) {
 	// use c.pkey to sign msg
 	return cryptopasta.Sign(msg, client.prikey.(*ecdsa.PrivateKey))
 }
+
+// Route gets route
+func (client *Client) Route(ctx context.Context) (route pb.PieceStoreRoutesClient, err error) {
+	if client.route == nil && client.conn == nil {
```
Comment (kaloyan-raev) [inline pkg/piecestore/rpc/client/client.go]: Perhaps just check `if client.route == nil`. It's valid for the `client.conn` to be nil if the PSClient was created with `NewCustomRoute`. The latter is used by unit tests which use mock route.

File: pkg/piecestore/rpc/client/pieceranger.go [hunk 3]
```diff
@@ -67,10 +67,22 @@ func (r *pieceRanger) Range(ctx context.Context, offset, length int64) (io.ReadC
 		return ioutil.NopCloser(bytes.NewReader([]byte{})), nil
 	}
 
+	stream := r.stream
```
Comment (kaloyan-raev) [inline pkg/piecestore/rpc/client/pieceranger.go]: What's the point of this helper variable. Can't you do all the below logic referring directly to `r.stream`?

File: examples/piecestore-client/rpc/client/main.go [hunk 4]
```diff
@@ -37,20 +38,23 @@ func main() {
...
-	}
-	defer conn.Close()
-
-	psClient, err := client.NewPSClient(conn, 1024*32, identity.Key.(*ecdsa.PrivateKey))
+	// var conn *grpc.ClientConn
+	// conn, err = grpc.Dial(":7777", identOpt)
+	// if err != nil {
+	// 	log.Fatalf("did not connect: %s", err)
+	// }
+	// defer conn.Close()
```
Comment (kaloyan-raev) [inline examples/piecestore-client/rpc/client/main.go]: You should remove all the commented code above.

File: pkg/piecestore/rpc/client/client.go [hunk 5]
```diff
@@ -95,12 +99,20 @@ func (client *Client) Close() error {
```
Comment (jtolds) [inline pkg/piecestore/rpc/client/client.go]: Close probably needs to get updated to handle a client that has a nil conn

**Comment by kaloyan-raev** [comment]
Yeah, @mniewrzal the approach with a lazy ranger that @jtolds proposed seems much simpler than that direction I gave you. This way you would not need to attach the PSClient and adapt its tests, which may be a little challenge too.

**Comment by mniewrzal** [comment]
Yes, @jtolds idea is simpler. I already tried this and it works nice but one way or another I need to adjust tests. I will push PR with the second approach to compare soon.

**Comment by kaloyan-raev** [comment]
Closing in favor of #321

**Comment by jtolds** [comment]
okay, in light of new issues, i think this approach is better after all. i'm going to rebase it on top of a revert of the other one

**Comment by jtolds** [comment]
TBD, we'll talk about this next week

**Comment by kaloyan-raev** [comment]
Closing. I believe we resolved issues with #343 and #349.

**Pull request closed without merging**